src/main/java/org/ossgang/commons/monads/AsyncMaybe.java

```diff
@@ -0,0 +1,284 @@
+     *
+     * @param supplier the function to apply
+     * @return A successful Maybe wrapping the return value if the function is executed and succeeds. An unsuccessful
+     * Maybe otherwise.
+     */
+    public <R> AsyncMaybe<R> then(ThrowingSupplier<R> supplier) {
+        requireNonNull(supplier);
+        return AsyncMaybe.fromCompletableFuture(future.thenApply(v -> unchecked(supplier).get()));
+    }
+
+    /**
+     * Apply a void function if this {@link Maybe} is in a "successful" state. Pass through the exception in case it is
+     * in an "unsuccessful" state. The function does not get the value of the Maybe passed as parameter, which is useful
+     * e.g. for chaining Maybe&lt;Void&gt;. If the function succeeds, a successful empty Maybe&lt;Void&gt; is returned.
+     *
+     * @param runnable the function to apply
+     * @return A successful Maybe&lt;Void&gt; it the function is executed and succeeds. An unsuccessful Maybe otherwise.
+     */
+    public AsyncMaybe<Void> then(ThrowingRunnable runnable) {
+        requireNonNull(runnable);
+        return AsyncMaybe.fromVoidCompletableFuture(future.thenRun(unchecked(runnable)));
+    }
+
+    /**
+     * Apply a function if this {@link Maybe} is in a "unsuccessful" state. The function gets passed the exception
+     * wrapped in this maybe. If it returns successfully, it's result is wrapped in a "successful" Maybe. If it
+     * throws, the exception is returned wrapped in an "unsuccessful" Maybe.
+     *
+     * @param function the recovery function to apply
+     * @return A successful Maybe the new value if the function returns. An unsuccessful Maybe if it throws.
+     */
+    public AsyncMaybe<T> recover(ThrowingFunction<Throwable, T> function) {
+        requireNonNull(function);
+        return AsyncMaybe.fromCompletableFuture(future.handle((value, exception) -> {
+            if (exception != null) {
+                return unchecked(function).apply(exception);
+            }
+            return value;
+        }));
+    }
+
+    /**
+     * Apply a function if this {@link Maybe} is in a "unsuccessful" state. The function gets passed the exception
+     * wrapped in this maybe. If it returns successfully, a "successful" void Maybe is returned. If it
+     * throws, the exception is returned wrapped in an "unsuccessful" Maybe.
+     *
+     * @param consumer the recovery handler to apply
+     * @return A successful void Maybe the new value if the handler returns. An unsuccessful Maybe if it throws.
+     */
+    public AsyncMaybe<Void> recover(ThrowingConsumer<Throwable> consumer) {
+        requireNonNull(consumer);
+        return AsyncMaybe.fromVoidCompletableFuture(future.handle((value, exception) -> {
+            if (exception != null) {
+                unchecked(consumer).accept(exception);
+            }
+            return null;
+        }));
+    }
+
+    /**
+     * TODO andrea
+     *
+     * @return
+     */
+    public Maybe<T> toMaybeBlocking() {
```

Comment:
perhaps we should have another method that allows to get with a timeout?
toMaybe() and toMaybe(Duration)?

Reply:
agree for the timeout :smile: 
I have misfeelings about letting go the blocking suffix.... I think it is a *really* important detail that may be overlooked. On the other hand.. you should know that a Maybe is not async and therefore this method has to block (and it will be written in the doc)